Pal 847 palisade root docs (#489)
gchq/Palisade: doc/ways_of_working.md

<!---
Copyright 2018-2021 Crown Copyright

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
--->

# Ways of Working

## Contents
1. [Git branching model](#git-branching-model)
1. [Issues](#issues)
1. [Pull requests](#pull-requests)
1. [Coding style](#coding-style)
1. [Javadoc](#javadoc)
1. [Tests](#tests)

## Git branching model
We have adopted the following branching strategy - see [Git Branching Model](https://nvie.com/files/Git-branching-model.pdf)

## Issues 

> Where possible a pull request should correlate to a single GitHub issue.

Reviewer: ```suggestion
Where possible, a pull request should correlate to a single GitHub issue.
```

An issue should relate to a single functional or non-functional change - changes to alter/improve other pieces of functionality should be addressed in a separate issue in order to keep reviews atomic.
The reasoning behind code changes should be documented in the GitHub issue. 
All resolved issues should be included in the next GitHub milestone, this enables releases to be linked to the included issues.

> If a code change requires users to make changes in order for them to adopt it then the issue should be labelled 'migration-required' and a comment should be added similar to:

Reviewer: ```suggestion
If a code change requires users to make changes in order for them to adopt it then the issue should be labelled 'migration-required', and a comment should be added similar to:
```

```
### Migration Steps

[Description of what needs to be done to adopt the code change with examples]
```

#### Workflow
* Assign yourself to the issue
* Create a new branch off develop using pattern `PAL-[issue number]-[issue-title]`
* Commit your changes prefixing your commit title with `PAL-[issue-number] [commit title]`
* Check your changes
* Create a pull request to merge your branch into develop
* The pull request will be reviewed and following any changes and approval your branch will be merged into develop
* Delete the branch

## Pull Requests
Pull requests will undergo an in depth review by a Palisade committer to check the code changes are compliant with our coding style.
This is a community so please be respectful of other members - offer encouragement, support and suggestions. 

> As described in our git branching model - please raise pull requests to merge you changes in our **develop** branch.

Reviewer: ```suggestion
As described in our git branching model - please raise pull requests to merge your changes into our **develop** branch.
```

  * This should apply to all volatile resource usage
* Don't swallow exceptions - ensure they are logged or rethrown
* Give credit for other peoples work
* Update the NOTICES file for changes to the dependencies
* Consider the scope of dependencies - restrict them when possible using the appropriate maven scope
* Don't expose private logic in classes through public methods
* Field access should be controlled via getters and setters
* Make use of the core Java API - don't reinvent the wheel
* Make use of generic typing
* Make use of appropriate object-oriented design patterns
* Use Loggers instead of System.out.print and Throwable.printStackTrace
* Ensure that the toString(), equals() and hashCode() methods are implemented where appropriate

#### Javadoc
Ensure your java code has sufficient Javadoc explaining what the section of code does and the intended use of it.
Javadoc should be used in addition to clean readable code, it is not an excuse to write lazy code.
Method bodies should also be accompanied by inline comments where appropriate.

In particular, the following are required:
* public classes (not required for test classes unless an explanation of the testing is required)

> * public methods (not required if the functionality is obvious from the method name e.g. getters and setters)

Reviewer 2: worth mentioning of the @Generated annotation?
